lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,429 @@
+// TODO: remove this when we integrate the new immutable id tracker
+#![allow(dead_code)]
+
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::{BufReader, BufWriter};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use bitvec::prelude::BitSlice;
+use bitvec::vec::BitVec;
+use common::types::PointOffsetType;
+use parking_lot::RwLock;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use crate::common::operation_error::{OperationError, OperationResult};
+use crate::common::Flusher;
+use crate::id_tracker::simple_id_tracker::StoredPointId;
+use crate::id_tracker::{IdTracker, IdTrackerEnum};
+use crate::types::{PointIdType, SeqNumberType};
+
+pub const DELETED_FILE_NAME: &str = "id_tracker.deleted";
+pub const MAPPINGS_FILE_NAME: &str = "id_tracker.mappings";
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct ImmutableIdTracker {
+    #[serde(skip)]
+    path: PathBuf,
+
+    deleted: BitVec,
+    dirty: Arc<RwLock<bool>>,
+
+    mappings: PointMappings,
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
+pub(super) struct PointMappings {
+    pub(crate) internal_to_version: Vec<SeqNumberType>,
+
+    pub(crate) internal_to_external: Vec<StoredPointId>,
+    pub(crate) external_to_internal: BTreeMap<StoredPointId, PointOffsetType>,
+}
+
+impl ImmutableIdTracker {
+    pub fn open(segment_path: &Path) -> OperationResult<Self> {
+        let deleted: BitVec = Self::open_file(segment_path.join(DELETED_FILE_NAME).as_path())?;
+        let mappings: PointMappings =
+            Self::open_file(segment_path.join(MAPPINGS_FILE_NAME).as_path())?;
+
+        Ok(Self {
+            path: segment_path.to_path_buf(),
+            deleted,
+            dirty: Arc::new(RwLock::new(false)),
+            mappings,
+        })
+    }
+
+    pub(super) fn new(path: PathBuf, deleted: BitVec, mappings: PointMappings) -> Self {
+        Self {
+            path,
+            deleted,
+            dirty: Arc::new(RwLock::new(false)),
+            mappings,
+        }
+    }
+
+    pub(super) fn save(&self) -> OperationResult<()> {
+        Self::save_file(&self.deleted_file_path(), &self.deleted)?;
+        Self::save_file(&self.mappings_file_path(), &self.mappings)?;
+        *self.dirty.write() = false;
+        Ok(())
+    }
```

Comment:
Is there a specific reason we need this?

I expect we want to save once when creating this immutable structure. But then I'd prefer to use the `::from(other)` approach, automatically saving once, keeping it internal only.